src/frequenz/core/enum.py

```diff
@@ -0,0 +1,204 @@
+    - The deprecated names are recorded so that:
+
+        * `MyEnum.NAME` warns (attribute access by name)
+        * `MyEnum["NAME"]` warns (lookup by name)
+        * `MyEnum(value)` warns **only if** the resolved member has **no**
+            non-deprecated aliases (all names for that member are deprecated).
+    """
+
+    def __new__(  # pylint: disable=too-many-locals
+        mcs,
+        name: str,
+        bases: tuple[type[EnumT], ...],
+        classdict: Mapping[str, Any],
+        **kw: Any,
+    ) -> type[EnumT]:
+        """Create the new enum class, rewriting `DeprecatedMember` instances."""
+        deprecated_names: dict[str, str] = {}
+        prepared = super().__prepare__(name, bases, **kw)
+
+        # Rewrite and record
+        for key, value in classdict.items():
+            if isinstance(value, DeprecatedMember):
+                deprecated_names[key] = value.message
+                prepared[key] = value.value
+            else:
+                prepared[key] = value
+
+        cls = cast(type[EnumT], super().__new__(mcs, name, bases, prepared, **kw))
+
+        # Build alias groups: member -> list of names
+        member_to_names: dict[EnumT, list[str]] = {}
+        member: EnumT
+        for member_name, member in cls.__members__.items():
+            member_to_names.setdefault(member, []).append(member_name)
+
+        warned_by_member: dict[EnumT, str] = {}
+        for member, names in member_to_names.items():
+            # warn on value only if all alias names are deprecated
+            deprecated_aliases = [n for n in names if n in deprecated_names]
+            if deprecated_aliases and len(deprecated_aliases) == len(names):
+                warned_by_member[member] = deprecated_names[deprecated_aliases[0]]
+
+        # Inject maps quietly
+        type.__setattr__(cls, "__deprecated_names__", deprecated_names)
+        type.__setattr__(cls, "__deprecated_value_map__", warned_by_member)
+
+        return cls
+
+    @staticmethod
+    def _name_map(cls_: type[Any]) -> Mapping[str, str]:
+        """Map from member names to deprecation messages."""
+        return cast(
+            Mapping[str, str],
+            type.__getattribute__(cls_, "__dict__").get("__deprecated_names__", {}),
+        )
+
+    @staticmethod
+    def _value_map(cls_: type[Any]) -> Mapping[Any, str]:
+        """Map from enum members to deprecation messages."""
+        return cast(
+            Mapping[Any, str],
+            type.__getattribute__(cls_, "__dict__").get("__deprecated_value_map__", {}),
+        )
+
+    def __getattribute__(cls, name: str) -> Any:
+        """Resolve `name` to a member, warning if the member is deprecated."""
+        if name in ("__deprecated_names__", "__deprecated_value_map__"):
+            return type.__getattribute__(cls, name)
+        # Metaclasses are weird, here mypy thinks cls is the metaclass but when this is
+        # actually used, it is the concrete type (E). We can't
```

Comment:
We can't .. what? don't leave me hanging

Reply:
Fixed in fixup commit